libnmstate/nm/nmclient.py

```diff
@@ -220,3 +234,53 @@ def _register_first_action(self):
     def _execute_action_once(self, _):
         self.execute_next_action()
         return False
+
+
+class _NmClient:
+    def __init__(self):
+        self.client = NM.Client.new(None)
+        logging.debug("NM.Client created")
+        self._mainloop = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, _exc_type, _exc_value, _traceback):
+        pass
+
+    def __del__(self):
+        try:
+            ref_count = sys.getrefcount(self.client)
+            if ref_count != 2:
+                raise error.NmstateInternalError(
+                    "_NmClient.__del__() with self.client holding "
+                    f"unexpected(not 2) reference count {ref_count}"
+                )
+            if NM.Client.get_version(self.client) >= StrictVersion("1.21.0"):
+                self._nmclient_cleanup()
+            else:
+                self._nmclient_cleanup_old()
+        except Exception:
+            pass
+
+    def _nmclient_cleanup_old(self):
+        self.client = None
+
+    def _nmclient_cleanup(self):
+        if self.client:
+            self._mainloop = _MainLoop()
+            context_busy_watcher = self.client.get_context_busy_watcher()
+            self._mainloop.push_action(
+                self._wait_nmclient_unref, context_busy_watcher
+            )
+            self.client = None
+            del context_busy_watcher
+            self._mainloop.run(_NMCLIENT_CLEANUP_TIMEOUT)
+
+    def _wait_nmclient_unref(self, context_busy_watcher):
+        context_busy_watcher.weak_ref(self._busy_watcher_destruct_callback)
+
+    def _busy_watcher_destruct_callback(self):
+        self._mainloop.execute_next_action()
+        self._mainloop = None
+        logging.debug("NM.Client cleaned")
```

Comment:
Seems to me like this should be called after `run()` returns.

And then, you could avoid most of these methods:
```
context_busy_watcher = self.client.get_context_busy_watcher()
mainloop = _MainLoop()
mainloop.push_action(
    lambda: context_busy_watcher.weak_ref(mainloop.execute_next_action)
)
mainloop.run(_NMCLIENT_CLEANUP_TIMEOUT)
logging.debug("NM.Client cleaned")
```

Reply:
With these lines, at the time of `mainloop.run()`, the `context_busy_watcher` is None. The mainloop should take a reference of it via `push_action().`

Reply:
Seems that the problem is with the release of `context_busy_watcher` before running the the mainloop. Same for the `client`.

Fix:
```
get_context_busy_watcher = self.client.get_context_busy_watcher
mainloop = _MainLoop()
mainloop.push_action(
    lambda: get_context_busy_watcher().weak_ref(mainloop.execute_next_action)
)
self._client = None
mainloop.run(_NMCLIENT_CLEANUP_TIMEOUT)
logging.debug("NM.Client cleaned")
```

Instead f using `del` or assigning `None`, we can try (when possible) to use the (block) scope which implicitly de-references the objects.